### PR TITLE
fix(controller): support float for param value

### LINKED ISSUE #4478: Workflow cannot read float
## Summary 

When a workflow parameter is a float, workflow parsing fails with the error message:
`Failed to parse workflow: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal number 0.2 into Go struct field Parameter.spec.arguments.parameters.value of type int`

What happened/what you expected to happen?
Would expect that the workflow is parsed correctly, float into type float.

## Diagnostics

What Kubernetes provider are you using? 
kube_version: 1.19.3 > running in VMs as local installation

when i run `kubectl version`  :
```bash
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:41:49Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```

What version of Argo Workflows are you running? 
```bash
argo: v2.11.7
  BuildDate: 2020-11-02T20:59:01Z
  GitCommit: bf3fec176cf6bdf3e23b2cb73ec7d4e3d051ca40
  GitTreeState: clean
  GitTag: v2.11.7
  GoVersion: go1.13.15
  Compiler: gc
  Platform: linux/amd64
```
```yaml
apiVersion: argoproj.io/v1alpha
kind: Workflow
metadata:
  generateName: argo-test-workflow-
spec:
  entrypoint: param-test
  arguments:
    parameters:

    # test parameters
    - name: STRING
      value: this is a string
    - name: INTEGER
      value: 5
    - name: FLOAT
      value: 0.2

  templates:
  - name: param-test
    container:
      image: alpine:latest
      command: [/bin/sh, -c]
      args: ["echo 'The parameter STRING is {{workflow.parameters.STRING}}.' && echo 'The parameter INTEGER is {{workflow.parameters.INTEGER}}' && echo 'The parameter FLOAT is {{workflow.parameters.FLOAT}}'"]
      env:
      - name: STRING
        value: "{{workflow.parameters.STRING}}"
      - name: INTEGER
        value: "{{workflow.parameters.INTEGER}}"
      - name: FLOAT
        value: "{{workflow.parameters.FLOAT}}"
```

```
Failed to parse workflow: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal number 0.2 into Go struct field Parameter.spec.arguments.parameters.value of type int

```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by arghya88**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Fixes #4478

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/int64orstr.go [hunk 1]
```diff
@@ -33,12 +33,12 @@ func (i *Int64OrString) UnmarshalJSON(value []byte) error {
...
 		return nil
 	}
-	v := 0
+	v := 0.0
 	err := json.Unmarshal(value, &v)
 	if err != nil {
 		return err
 	}
-	*i = Int64OrString(strconv.Itoa(v))
+	*i = Int64OrString(strconv.FormatFloat(v, 'f', -1, 64))
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/int64orstr.go:41]: could we rename this type `AnyString` and add support for `bool`?

Comment (arghya88) [reply pkg/apis/workflow/v1alpha1/int64orstr.go]: Done

**Comment by alexec** [comment]
You have a failing unit test?

**Comment by arghya88** [comment]
> You have a failing unit test?

@alexec probably flake. Fixed now

File: pkg/apis/workflow/v1alpha1/anystring.go [hunk 2]
```diff
@@ -12,40 +12,45 @@ import (
...
 		return nil
 	}
-	v := 0
+	var v interface{}
 	err := json.Unmarshal(value, &v)
 	if err != nil {
 		return err
 	}
-	*i = Int64OrString(strconv.Itoa(v))
+	switch v := v.(type) {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/anystring.go:41]: could this switch handle the string case too?

File: pkg/apis/workflow/v1alpha1/anystring.go [hunk 3]
```diff
@@ -12,40 +12,45 @@ import (
...
+	case float64:
+		*i = AnyString(strconv.FormatFloat(v, 'f', -1, 64))
```
Comment (terrytangyuan) [inline pkg/apis/workflow/v1alpha1/anystring.go:43]: What about float32?

File: pkg/apis/workflow/v1alpha1/anystring.go [hunk 4]
```diff
@@ -12,40 +12,45 @@ import (
 // * It's JSON type is just string, not `int-or-string`.
 // * It will unmarshall int64 (rather than only int32) and represents it as string.
```
Comment (terrytangyuan) [inline pkg/apis/workflow/v1alpha1/anystring.go:13]: The comments here need to be updated as well

File: pkg/apis/workflow/v1alpha1/anystring_test.go [hunk 5]
```diff
@@ -7,57 +7,83 @@ import (
...
+		if assert.NoError(t, err) {
+			assert.Equal(t, AnyStringPtr("0.2"), i)
+		}
+		assert.Equal(t, "0.2", i.String())
+	})
+	t.Run("Boolean", func(t *testing.T) {
+		x := ParseAnyString(true)
+		data, err := json.Marshal(x)
+		if assert.NoError(t, err) {
+			assert.Equal(t, `"true"`, string(data))
```
Comment (terrytangyuan) [inline pkg/apis/workflow/v1alpha1/anystring_test.go:80]: Perhaps add a test case for `false`  here as well

**Comment by terrytangyuan** [comment]
I was just about the send a similar PR to fix this as well and then found this one. Thanks @arghya88!

**Comment by alexec** [comment]
@terrytangyuan good comment - thank you!

File: pkg/apis/workflow/v1alpha1/anystring.go (hunk 2 above)
Comment (arghya88) [reply pkg/apis/workflow/v1alpha1/anystring.go:29]: Seems so.

**Review by terrytangyuan — APPROVED** [review]
Thanks @arghya88 the changes look great! @alexec PTAL

**Review by alexec — APPROVED** [review]
We should back-port to v2.12 if easy. v3 if not.

File: pkg/apis/workflow/v1alpha1/anystring.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+func (i *AnyString) UnmarshalJSON(value []byte) error {
+	var v interface{}
+	err := json.Unmarshal(value, &v)
+	if err != nil {
+		return err
+	}
+	switch v := v.(type) {
+	case float64:
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/anystring.go:30]: what about int64?

**Comment by alexec** [comment]
Needs backporting to v2.12

File: pkg/apis/workflow/v1alpha1/anystring.go (hunk 6 above)
Comment (arghya88) [reply pkg/apis/workflow/v1alpha1/anystring.go:30]: Added

**Comment by arghya88** [comment]
> Needs backporting to v2.12

@alexec I will submit a PR for that

**Comment by alexec** [comment]
> > Needs backporting to v2.12
> 
> @alexec I will submit a PR for that

No need. Will do this for you.

**Pull request merged**